### ISSUE TITLE
Using multiple statefulsets for each component

### ISSUE THREAD

**Issue by tennix** [pingcap/tidb-operator#228]
Currently, each component has one statefulset, and the statefulset controls the component's behavior. This works well for normal cases. But have some limitations:
* When using local volume, only horizontal scaling is guaranteed, vertical scaling may fail if there are not enough resources on the node.
* Users can't specify an exact number of pods in a given AZ. For example deploy 3 TiKV in AZ-1, and 2 TiKV in AZ-2.
* For on-premise deployment, k8s nodes may not have the same size. To make most of the resources, it's better to allow different profiles for a single cluster.
* No isolation between different tidb-servers, a big AP query may affect TP query. The worst case is that AP query cause a tidb-server OOM, and this tidb-server is also serving TP request.

When switched to multiple statefulsets, all the above problems can be resolved.

**Comment by gregwebs** [comment]
The components in question are just PD and TiKV?

> For on-premise deployment, k8s nodes may not have the same size. To make most of the resources, it's better to allow different profiles for a single cluster.

Is this resolvable by using the various resource requests parameters to allow specifying a range?

> Users can't specify an exact number of pods in a given AZ. For example deploy 3 TiKV in AZ-1, and 2 TiKV in AZ-2.

Is this resolvable by telling the operator this layout and having the operator interact with the scheduler? Essentially by default make TiKV not deployable and then change labels to get them deployed where we want?


In my existing thinking about multiple stateful sets, it was a temporary condition to allow a scaling transition. However, it sounds like this is leaning closer to having a separate stateful set per pod? I think what we need is just a StatefulPod and then have group behavior manage by our operator. So it seems like a StatefulSet could become our StatefulPod? Its also worth considering researching how others have tackled this challenge and seeing if we can come up with a better abstraction.

**Comment by tennix** [comment]
No, the components include PD/TiKV/TiDB. See the last issue I listed.

The resource limits and requests may both different in a single cluster for on-premise deployment.

It is doable by hacking the scheduler extender to deploy an exact number of pods in a given AZ. But considering the other issues I listed, multiple statefulsets would be a more elegant approach.

By using multiple statefulsets, we're not making it separate statefulset per pod. Though it's possible if users deploy like that. We're still relying on StatefulSet controller to manage pods. Pods are still grouped as a statefulset, but are split as we needed (Geo-location, Heterogeneous nodes, Isolation workloads).
